src/workerd/jsg/url.c++

```diff
@@ -0,0 +1,240 @@
+#include "url.h"
+
+extern "C" {
+  #include <ada_c.h>
+}
+
+#include <kj/debug.h>
+
+namespace workerd::jsg {
+
+namespace {
+class AdaUrlDisposer : public kj::Disposer {
+public:
+  static const AdaUrlDisposer INSTANCE;
+
+protected:
+  void disposeImpl(void* pointer) const override {
+    ada_free(pointer);
+  }
+};
+const AdaUrlDisposer AdaUrlDisposer::INSTANCE;
+
+class AdaOwnedStringDisposer : public kj::ArrayDisposer {
+public:
+  static const AdaOwnedStringDisposer INSTANCE;
+
+protected:
+  void disposeImpl(void* firstElement, size_t elementSize, size_t elementCount,
+                   size_t capacity, void (*destroyElement)(void*)) const {
+    ada_owned_string data = {
+      static_cast<const char*>(firstElement),
+      elementCount };
+    ada_free_owned_string(data);
+  }
+};
+const AdaOwnedStringDisposer AdaOwnedStringDisposer::INSTANCE;
+
+kj::Own<void> wrap(ada_url url) {
+  return kj::Own<void>(url, AdaUrlDisposer::INSTANCE);
+}
+
+}  // namespace
+
+Url::Url(kj::Own<void> inner) : inner(kj::mv(inner)) {}
+
+Url::Url(Url&& other) : inner(kj::mv(other.inner)) {}
+
+Url& Url::operator=(Url&& other) {
+  inner = kj::mv(other.inner);
+  return *this;
+}
+
+bool Url::operator==(const Url& other) const {
+  return getHref() == other.getHref();
+}
+
+bool Url::canParse(kj::StringPtr input, kj::Maybe<kj::StringPtr> base) {
+  KJ_IF_SOME(b, base) {
+    return ada_can_parse_with_base(input.begin(), input.size(),
+                                   b.begin(), b.size());
+  }
+  return ada_can_parse(input.begin(), input.size());
+}
+
+kj::Maybe<Url> Url::tryParse(kj::ArrayPtr<const char> input,
+                             kj::Maybe<kj::ArrayPtr<const char>> base) {
+  ada_url result = nullptr;
+  KJ_IF_SOME(b, base) {
+    result = ada_parse_with_base(input.begin(), input.size(),
+                                 b.begin(), b.size());
+  } else {
+    result = ada_parse(input.begin(), input.size());
+  }
+  if (!ada_is_valid(result)) return kj::none;
+  return Url(wrap(result));
+}
+
+kj::Maybe<Url> Url::resolve(kj::ArrayPtr<const char> input) {
+  return tryParse(input, getHref());
+}
+
+kj::ArrayPtr<const char> Url::getHref() const {
+  ada_string href = ada_get_href(const_cast<void*>(inner.get()));
+  return kj::ArrayPtr<const char>(href.data, href.length);
+}
+
+kj::Maybe<kj::ArrayPtr<const char>> Url::getUsername() const {
+  if (ada_has_credentials(const_cast<void*>(inner.get()))) {
```

Comment:
It looks like `ada_has_credentials` returns true if either, or both, username and password are present.

Could this method return ada_get_username unconditionally?

Or should it gate with a check that there is a non-empty username?
[ada_has_non_empty_username](https://cpp-api.ada-url.com/dd/dd6/ada__c_8cpp.html#ae6b8b22187c2379cd9331d16a6cd4f8f)

No opinion here, just looking at the ada docs.

Reply:
According to spec. username needs to exist in order to have a password. Regardless, we can just get username by ada_get_username and return None if length is 0. Might be a faster approach

Reply:
I think we can likely remove the `kj::Maybe` on all of these getters. The implementation (and the spec) returns these as zero-length strings anyway if they are not specified.